File: frontend/js/views/debt_yield.js
// Debt yield & loan sizing — commercial-RE lender ratios (debt yield, LTV,
// LTC) and the max loan each allows, with the binding constraint, via
// /calc/debt-yield. Updates live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const FIELDS = [
    ['noi_usd', 'Net operating income ($)', 100000],
    ['property_value_usd', 'Appraised value ($)', 1400000],
    ['total_project_cost_usd', 'Total project cost ($)', 1350000],
    ['loan_amount_usd', 'Proposed loan ($)', 1000000],
    ['min_debt_yield_pct', 'Min debt yield (%)', 10],
    ['max_ltv_pct', 'Max LTV (%)', 75],
    ['max_ltc_pct', 'Max LTC (%)', 80],
];

const money = (n) => '$' + Number(n).toLocaleString(undefined, { maximumFractionDigits: 0 });
const pct = (n) => Number(n).toFixed(2) + '%';
const BINDING = { debt_yield: 'Debt yield', ltv: 'LTV', ltc: 'LTC' };
const VIEW = 'debt-yield';
let lastReport = null;
let lastBody = null;

export async function renderDebtYield(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.dy.h1.title">// DEBT YIELD & LOAN SIZING</span></h1>
        <p class="muted small" data-i18n="view.dy.hint.intro">
            A commercial lender sizes a loan against three ceilings, and the smallest wins:
            debt yield (NOI ÷ loan, a rate-independent risk measure, floored at ~8–10%), LTV
            (loan ÷ value), and LTC (loan ÷ total cost). This shows each ratio for your proposed
            loan, the max loan each ceiling allows, and which one binds. Complements cap rate
            and DSCR. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.dy.h2.inputs">The deal</h2>
            <form id="dy-form" class="inline-form">
                ${FIELDS.map(([key, label, def]) => `
                    <label><span data-i18n="view.dy.label.${key}">${label}</span>
                        <input type="number" step="0.01" min="0" name="${key}" value="${def}" required></label>
                `).join('')}
            </form>
            <div id="dy-tools" class="ce-toolbar"></div>
            <button type="button" id="dy-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="dy-sens" class="ce-sens"></div>
        </div>
        <div id="dy-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#dy-form');
    enh.prefillForm(form, enh.readHashInputs());
    const readBody = () => {
        const fd = new FormData(form);
        const body = {};
        for (const [key] of FIELDS) body[key] = Number(fd.get(key)) || 0;
        return body;
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcDebtYield(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.dy.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#dy-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'debt-yield.csv' });
    mount.querySelector('#dy-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['max_loan_usd', r.max_loan_usd],
        ['binding_constraint', r.binding_constraint],
        ['loan_fits', r.loan_fits],
        ['debt_yield_pct', r.debt_yield_pct],
        ['ltv_pct', r.ltv_pct],
        ['ltc_pct', r.ltc_pct],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#dy-result');
    const fitCls = r.loan_fits ? 'pos' : 'neg';
    const binding = BINDING[r.binding_constraint] || r.binding_constraint;
    // Line chart: max loan as min debt yield sweeps 8% → 14% (loan ceiling falls as the floor rises).
    const xs = enh.linspace(8, 14, 13);
    const pts = await Promise.all(xs.map(async (dy) => {
        const rr = await api.calcDebtYield({ ...body, min_debt_yield_pct: dy });
        return { x: dy, y: rr ? rr.max_loan_usd / 1000 : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'min DY %', ylabel: 'max loan $k' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.dy.h2.result">Loan sizing</h2>
            <div class="cards">
                <div class="card ${fitCls}"><div class="label" data-i18n="view.dy.card.maxloan">Max loan</div>
                    <div class="value ${fitCls}">${money(r.max_loan_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.dy.card.binding">Binding constraint</div>
                    <div class="value">${binding}</div></div>
                <div class="card ${fitCls}"><div class="label" data-i18n="view.dy.card.fits">Proposed loan fits?</div>
                    <div class="value ${fitCls}">${r.loan_fits ? t('view.dy.yes') : t('view.dy.no')}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <thead><tr>
                    <th data-i18n="view.dy.col.metric">Metric</th>
                    <th data-i18n="view.dy.col.current">Current</th>
                    <th data-i18n="view.dy.col.maxloan">Max loan</th>
                </tr></thead>
                <tbody>
                    <tr><td data-i18n="view.dy.row.dy">Debt yield</td><td>${pct(r.debt_yield_pct)}</td><td>${money(r.max_loan_by_debt_yield_usd)}</td></tr>
                    <tr><td data-i18n="view.dy.row.ltv">LTV</td><td>${pct(r.ltv_pct)}</td><td>${money(r.max_loan_by_ltv_usd)}</td></tr>
                    <tr><td data-i18n="view.dy.row.ltc">LTC</td><td>${pct(r.ltc_pct)}</td><td>${money(r.max_loan_by_ltc_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#dy-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: min debt yield 8% → 14%; y: max LTV 60% → 85%. Output: max loan.
    const xVals = enh.linspace(8, 14, 5);
    const yVals = enh.linspace(60, 85, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'min_debt_yield_pct', yKey: 'max_ltv_pct', xVals, yVals, compute: (b) => api.calcDebtYield(b), pick: (r) => (r ? r.max_loan_usd : null) });
    if (!viewIsCurrent(tok)) return;
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells, fmt: (v) => (v == null ? '—' : '$' + Math.round(v / 1000) + 'k'), xfmt: (v) => v.toFixed(0) + '%', yfmt: (v) => v.toFixed(0) + '%', xName: t('view.dy.label.min_debt_yield_pct') || 'Min DY', yName: t('view.dy.label.max_ltv_pct') || 'Max LTV' });
}

File: crates/traderview-core/src/debt_yield.rs
//! Debt yield & loan sizing — the commercial-RE lender constraints.
//!
//! A lender sizes a commercial real-estate loan against three ceilings, and
//! the smallest wins:
//!
//!   * **Debt yield** = NOI / loan. A rate- and amortization-independent
//!     measure of risk; lenders want it above a floor (often 8–10%). Max loan
//!     = NOI / floor.
//!   * **LTV** = loan / appraised value. Max loan = value × max LTV.
//!   * **LTC** = loan / total project cost. Max loan = cost × max LTC.
//!
//! Reports each ratio for a proposed loan, the max loan each ceiling allows,
//! the binding constraint, and whether the proposed loan fits. Complements
//! cap rate (NOI/value) and DSCR (NOI/debt-service). Pure compute.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct DebtYieldInput {
    pub noi_usd: f64,
    pub property_value_usd: f64,
    /// Total project cost (purchase + rehab) for the LTC test.
    pub total_project_cost_usd: f64,
    /// Proposed loan amount to evaluate.
    pub loan_amount_usd: f64,
    /// Lender's minimum debt yield (e.g. 10).
    pub min_debt_yield_pct: f64,
    /// Lender's maximum LTV (e.g. 75).
    pub max_ltv_pct: f64,
    /// Lender's maximum LTC (e.g. 80).
    pub max_ltc_pct: f64,
}

#[derive(Debug, Clone, Serialize)]
pub struct DebtYieldResult {
    pub debt_yield_pct: f64,
    pub ltv_pct: f64,
    pub ltc_pct: f64,
    pub max_loan_by_debt_yield_usd: f64,
    pub max_loan_by_ltv_usd: f64,
    pub max_loan_by_ltc_usd: f64,
    /// The smallest of the three ceilings — the actual max loan.
    pub max_loan_usd: f64,
    /// Which ceiling binds: "debt_yield" / "ltv" / "ltc".
    pub binding_constraint: String,
    /// True when the proposed loan is within the max.
    pub loan_fits: bool,
}

pub fn analyze(i: &DebtYieldInput) -> DebtYieldResult {
    let loan = i.loan_amount_usd.max(0.0);

    let debt_yield = if loan > 0.0 { i.noi_usd / loan * 100.0 } else { 0.0 };
    let ltv = if i.property_value_usd > 0.0 { loan / i.property_value_usd * 100.0 } else { 0.0 };
    let ltc = if i.total_project_cost_usd > 0.0 { loan / i.total_project_cost_usd * 100.0 } else { 0.0 };

    let max_by_dy = if i.min_debt_yield_pct > 0.0 { i.noi_usd / (i.min_debt_yield_pct / 100.0) } else { f64::INFINITY };
    let max_by_ltv = i.property_value_usd * i.max_ltv_pct / 100.0;
    let max_by_ltc = i.total_project_cost_usd * i.max_ltc_pct / 100.0;

    // The binding ceiling is the smallest.
    let mut max_loan = max_by_dy;
    let mut binding = "debt_yield";
    if max_by_ltv < max_loan {
        max_loan = max_by_ltv;
        binding = "ltv";
    }
    if max_by_ltc < max_loan {
        max_loan = max_by_ltc;
        binding = "ltc";
    }

    DebtYieldResult {
        debt_yield_pct: debt_yield,
        ltv_pct: ltv,
        ltc_pct: ltc,
        max_loan_by_debt_yield_usd: if max_by_dy.is_finite() { max_by_dy } else { 0.0 },
        max_loan_by_ltv_usd: max_by_ltv,
        max_loan_by_ltc_usd: max_by_ltc,
        max_loan_usd: if max_loan.is_finite() { max_loan } else { 0.0 },
        binding_constraint: binding.to_string(),
        loan_fits: max_loan.is_finite() && loan <= max_loan,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn base() -> DebtYieldInput {
        DebtYieldInput {
            noi_usd: 100_000.0,
            property_value_usd: 1_400_000.0,
            total_project_cost_usd: 1_350_000.0,
            loan_amount_usd: 1_000_000.0,
            min_debt_yield_pct: 10.0,
            max_ltv_pct: 75.0,
            max_ltc_pct: 80.0,
        }
    }

    #[test]
    fn debt_yield_is_noi_over_loan() {
        // 100k / 1M = 10%.
        let r = analyze(&base());
        assert!((r.debt_yield_pct - 10.0).abs() < 1e-9);
    }

    #[test]
    fn ltv_and_ltc() {
        let r = analyze(&base());
        assert!((r.ltv_pct - 1_000_000.0 / 1_400_000.0 * 100.0).abs() < 1e-9);
        assert!((r.ltc_pct - 1_000_000.0 / 1_350_000.0 * 100.0).abs() < 1e-9);
    }

    #[test]
    fn max_loan_by_debt_yield() {
        // NOI 100k / 10% = 1,000,000.
        let r = analyze(&base());
        assert!((r.max_loan_by_debt_yield_usd - 1_000_000.0).abs() < 1e-6);
    }

    #[test]
    fn max_loan_by_ltv_and_ltc() {
        let r = analyze(&base());
        assert!((r.max_loan_by_ltv_usd - 1_050_000.0).abs() < 1e-6); // 1.4M × 75%
        assert!((r.max_loan_by_ltc_usd - 1_080_000.0).abs() < 1e-6); // 1.35M × 80%
    }

    #[test]
    fn binding_constraint_is_the_smallest() {
        // DY 1.0M < LTV 1.05M < LTC 1.08M → debt yield binds.
        let r = analyze(&base());
        assert_eq!(r.binding_constraint, "debt_yield");
        assert!((r.max_loan_usd - 1_000_000.0).abs() < 1e-6);
    }

    #[test]
    fn ltv_binds_when_value_is_low() {
        // Low value makes LTV ceiling the smallest.
        let r = analyze(&DebtYieldInput { property_value_usd: 1_000_000.0, ..base() });
        // LTV max = 750k, DY max = 1.0M, LTC max = 1.08M → LTV binds.
        assert_eq!(r.binding_constraint, "ltv");
        assert!((r.max_loan_usd - 750_000.0).abs() < 1e-6);
    }

    #[test]
    fn loan_fits_within_max() {
        // Proposed 1.0M ≤ max 1.0M → fits (boundary).
        assert!(analyze(&base()).loan_fits);
        // Over the max → doesn't fit.
        let over = analyze(&DebtYieldInput { loan_amount_usd: 1_200_000.0, ..base() });
        assert!(!over.loan_fits);
    }

    #[test]
    fn zero_loan_guards_ratios() {
        let r = analyze(&DebtYieldInput { loan_amount_usd: 0.0, ..base() });
        assert!(r.debt_yield_pct.abs() < 1e-9);
        assert!(r.ltv_pct.abs() < 1e-9);
        assert!(r.loan_fits); // 0 ≤ any max
    }
}
